Duplicate notebooks does not work

I'm using the current Master (build from spark-notebook.io , 0.8.0-SNAPSHOT-8d5304ca501e856fbda07b0a26a0cdc8b67b47a3 ) and having problems saving my notebook while using the default filesystem storage provider (as given in application.conf)

When I try to duplicate a Notebook (in my case `ClusterTest` which is top-level, i.e. under `notebooks/ClusterTest.snb`) does does nothing, I find in the logs:

```
17/05/03 11:49:58 ERROR play: Cannot invoke the action, eventually got an error: java.nio.file.NoSuchFileException: /home/myuser/spark-notebook-0.8.0-SNAPSHOT-scala-2.11.8-spark-2.1.1-hadoop-2.6.0-cdh5.8.4-with-hive/notebooks/../../notebooks/ClusterTest.snb
17/05/03 11:49:58 ERROR application:

! @73pgff1ho - Internal server error, for (POST) [/api/contents/] ->

@73pgff1ho: Execution exception in null:null
        at play.api.Application$class.handleError(Application.scala:296)
        at play.api.DefaultApplication.handleError(Application.scala:402)
        at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$3$$anonfun$applyOrElse$4.apply(PlayDefaultUpstreamHandler.scala:320)
        at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$3$$anonfun$applyOrElse$4.apply(PlayDefaultUpstreamHandler.scala:320)
        at scala.Option.map(Option.scala:146)
        at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$3.applyOrElse(PlayDefaultUpstreamHandler.scala:320)
        at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$3.applyOrElse(PlayDefaultUpstreamHandler.scala:316)
        at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:346)
        at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:345)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
        at play.api.libs.iteratee.Execution$trampoline$.execute(Execution.scala:46)
        at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:40)
        at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:248)
        at scala.concurrent.Promise$class.complete(Promise.scala:55)
        at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:153)
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:23)
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.nio.file.NoSuchFileException:  /home/myuser/spark-notebook-0.8.0-SNAPSHOT-scala-2.11.8-spark-2.1.1-hadoop-2.6.0-cdh5.8.4-with-hive/notebooks/../../notebooks/ClusterTest.snb
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
        at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
        at java.nio.file.Files.newByteChannel(Files.java:361)
        at java.nio.file.Files.newByteChannel(Files.java:407)
        at java.nio.file.Files.readAllBytes(Files.java:3152)
        at notebook.io.FileSystemNotebookProviderConfigurator$FileSystemNotebookProvider$$anonfun$get$1.apply(FileSystemNotebookProviderConfigurator.scala:46)
        at notebook.io.FileSystemNotebookProviderConfigurator$FileSystemNotebookProvider$$anonfun$get$1.apply(FileSystemNotebookProviderConfigurator.scala:46)
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
        at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
        ... 4 more

```

--- a/modules/core/src/main/scala/notebook/io/FileSystemNotebookProviderConfigurator.scala
+++ b/modules/core/src/main/scala/notebook/io/FileSystemNotebookProviderConfigurator.scala
@@ -15,7 +15,7 @@ class FileSystemNotebookProviderConfigurator extends Configurable[NotebookProvid
 
   override def apply(config: Config)(implicit ec:ExecutionContext): Future[NotebookProvider] = {
     val rootPath = Future {
-      Paths.get(config.getString(NotebooksDir))
+      getAbsoluteCanonicalPath(config.getString(NotebooksDir))
     }.recoverWith{ case t: Throwable =>
       Future.failed(new ConfigurationMissingException(NotebooksDir))
     }
@@ -67,4 +67,13 @@ class FileSystemNotebookProviderConfigurator extends Configurable[NotebookProvid
 }
 object FileSystemNotebookProviderConfigurator {
   val NotebooksDir = "notebooks.dir"
+
+  /**
+    * Get absolute path and remove any redundant ./ or ../../
+    */
+  def getAbsoluteCanonicalPath(path: String): Path = {
+    Paths.get(path)
+      .toAbsolutePath
+      .toFile.getCanonicalFile.toPath
+  }
 }

--- a/modules/core/src/test/scala/notebook/io/FileSystemNotebookProviderConfiguratorTests.scala
+++ b/modules/core/src/test/scala/notebook/io/FileSystemNotebookProviderConfiguratorTests.scala
@@ -15,11 +15,12 @@ class FileSystemNotebookProviderConfiguratorTests extends WordSpec with Matchers
   var notebookDir : String = _
   var tempDir: Path = _
 
+  val notebooksDirName = "notebook"
+
   override def beforeAll: Unit = {
     tempDir = Files.createTempDirectory("file-system-notebook-provider")
-    notebookDir = tempDir.toAbsolutePath.toFile.getAbsolutePath + "/notebook"
+    notebookDir = s"${tempDir.toAbsolutePath.toFile.getAbsolutePath}/$notebooksDirName"
   }
-
 
   "File system notebook provider configurator" should {
 
@@ -35,6 +36,20 @@ class FileSystemNotebookProviderConfiguratorTests extends WordSpec with Matchers
       val notebookProvider = configurator(dirConfig)
       whenReady(notebookProvider) { nbp =>
         nbp shouldBe a[NotebookProvider]
+      }
+    }
+
+    "accepts and cleans up non-cannonical relative paths" in {
+      val nonCanonicalNotebooksDir = notebookDir + s"/../$notebooksDirName"
+
+      val configurator = instantiateProvider()
+      val dirConfig = ConfigFactory.parseMap(Map("notebooks.dir" -> nonCanonicalNotebooksDir).asJava)
+      val notebookProvider = configurator(dirConfig)
+      whenReady(notebookProvider) { nbp =>
+        nbp shouldBe a[NotebookProvider]
+        // returns a root without notebook/../notebook
+        // (just ignore OSx tmpdir prefix '/private' for now)
+        nbp.root.toString.stripPrefix("/private") shouldBe notebookDir
       }
     }
 

